### PR TITLE
Fix `MissingResource` properties being stripped on save

### LINKED ISSUE #86599: Resources with properties set to missing resource types are stripped upon save
### Tested versions

Current master (13a0d6e9b253654f5cc2a44f3d0b3cae10440443)

### System information

MacOS - Godot v4.3-13a0d6e9

### Issue description

Resources that have properties set to instances of missing resource types (e.g. provided by an extension that isn't currently loaded) will have those properties and any missing-type sub/ext-resources stripped out upon save in the editor.

It's not supposed to do this; this is the entire reason that `MissingResource` was implemented (see #60597).

### Steps to reproduce

Using the MRP below:

Save FontFileExtResource in the inspector, open in a text editor. Notice it has removed the sub_resource and the property set to it.
Save MissingExtResourceTest, open in a text editor. Notice it has removed the external resource and property set to it.
Save MissingSubResourceTest, open in a text editor. Notice it has removed the sub_resource and the property set to it.

### Minimal reproduction project (MRP)

[test_missing_resources.zip](https://github.com/godotengine/godot/files/13791178/test_missing_resources.zip)

### PR DESCRIPTION
**Pull request by nikitalita**
fixes #86599

See https://github.com/godotengine/godot/pull/60597#pullrequestreview-1798576158

There were a number of bugs in the MissingResource implementation that prevented it from actually preserving the missing resources when saving; I have corrected these.

### CODE REVIEW

**Comment by fire** [comment]
@reduz Preserving the missing resources when saving is a good thing, but I would like a technical second look.

**Comment by fire** [comment]
Can you do a rebase? I'll try to review.

**Comment by nikitalita** [comment]
Thanks, I've rebased on current master

**Comment by nikitalita** [comment]
@fire I've rebased again, can you take a look?

**Review by fire — APPROVED** [review]
I think it's ok, but it's also touching core so I'm a bit wary.

Will try it on my projects.

File: scene/resources/resource_format_text.cpp [hunk 1]
```diff
@@ -723,24 +723,25 @@ Error ResourceLoaderText::load() {
...
 				}
-				return error;
+				break;
 			}
 
 			if (!assign.is_empty()) {
 				bool set_valid = true;
 
-				if (value.get_type() == Variant::OBJECT && missing_resource != nullptr) {
+				if (value.get_type() == Variant::OBJECT && ResourceLoader::is_creating_missing_resources_if_class_unavailable_enabled() && missing_resource == nullptr) {
```
Comment (AThousandShips) [inline scene/resources/resource_format_text.cpp:744]:
```suggestion
				if (value.get_type() == Variant::OBJECT && missing_resource == nullptr && ResourceLoader::is_creating_missing_resources_if_class_unavailable_enabled()) {
```

**Review by AThousandShips — APPROVED** [review]
Style looks good

File: core/io/missing_resource.h [hunk 2]
```diff
@@ -57,6 +57,8 @@ class MissingResource : public Resource {
 	void set_recording_properties(bool p_enable);
 	bool is_recording_properties() const;
 
+	virtual String get_save_class() const override;
```
[the remark this answers is not in the document]
Comment (nikitalita) [reply core/io/missing_resource.h:60]: `get_save_class()` is called by the resources savers when saving the class type to the resource. If it's a missing resource, we need to emit the original class type here or otherwise they will write the type as "MissingResource"

**Review by fire — APPROVED** [review]

**Comment by hpvb** [comment]
LGTM

**Comment by lyuma** [comment]
The bug and this PR are fixing an issue "stripped upon save". Why then is it modifying load-side functions ResourceLoader and scenestate::_parse_node?

i am concerned that this implies some form of compat breakage (resources changing next time I update the engine)

**Comment by nikitalita** [comment]
> The bug and this PR are fixing an issue "stripped upon save". Why then is it modifying load-side functions ResourceLoader and scenestate::_parse_node?
> 
> i am concerned that this implies some form of compat breakage (resources changing next time I update the engine)

Please see [here](https://github.com/godotengine/godot/pull/60597#pullrequestreview-1798576158) for a full explanation, but basically:

The reason for ResourceLoader being modified is that the pre-existing missing resource assignment was buggy; thus, the missing resources were not actually being loaded as MissingResources, and as such were not being saved when the scene/resource was re-saved.

The reason for scenestate::_parse_node being modified: 

> Nodes report their theme override properties as having `PROPERTY_USAGE_STORAGE` based on whether or not they're actually set. If a theme override property was set to a missing resource, it will be reported as not having `PROPERTY_USAGE_STORAGE` and it will be skipped. This should check if `missing_resource_properties` has the property.

Basically, when saving a scene, nodes that had properties set to a missing resource were skipped because they weren't previously loaded into the packed scene. 

This shouldn't mess with scene instancing. During actual instancing, when attempting to assign a MissingResource to a node property, it'll just silently fail.

**Comment by Repiteo** [comment]
Thanks!
